### ISSUE TITLE
CSINode object is not recreated after Node object is recreated

### ISSUE THREAD

**Issue by msau42** [kubernetes/kubernetes#120063]
### What happened?

A preemptible Node got deleted and recreated with the same name, however during csi driver registration the CSINode object was updated while it still had an ownerReference to the old Node uid. Then the CSINode object got garbage collected and we're left with no CSINode object, causing attach/mount operations to fail.

Some audit logs with a timeline of events:

Old node uuid is `ef26e636-556b-4af7-ba9e-fecece154997`

The replacement Node object was created after preemption:
```
DEFAULT 2023-07-28T07:59:59.649269Z [protoPayload.methodName: io.k8s.core.v1.nodes.create] 
        "metadata": {
          "uid": "bec5b072-7eb8-4ab9-a5c7-e4242414e3e0"
        },
```

There was an update to CSINode to register the driver, however the CSINode object updated has an owner reference to the old Node uid.
```
DEFAULT 2023-07-28T08:00:33.437252Z [protoPayload.methodName: io.k8s.storage.v1.csinodes.update] 
          "ownerReferences": [
            {
              "apiVersion": "v1",
              "kind": "Node",
              "uid": "ef26e636-556b-4af7-ba9e-fecece154997"
            }
        },
        "spec": {
          "drivers":  ....
       }
```

The CSINode was then garbage collected because the old Node object was deleted.
```
DEFAULT 2023-07-28T08:01:09.931137Z [protoPayload.methodName: io.k8s.storage.v1.csinodes.delete] [protoPayload.authenticationInfo.principalEmail: system:serviceaccount:kube-system:generic-garbage-collector] 
          "ownerReferences": [
            {
              "apiVersion": "v1",
              "kind": "Node",
              "uid": "ef26e636-556b-4af7-ba9e-fecece154997"
            }
```

Then CSINode was not recreated until some time later, after the driver was manually restarted.

### What did you expect to happen?

CSI driver registration should update/recreate the correct CSINode object.

### How can we reproduce it (as minimally and precisely as possible)?

It's a race condition between the garbage collector and csi driver registration.

The csi driver needs to start up and register before the old CSINode object gets garbage collected.

### Kubernetes version

<details>

```console
$ kubectl version
1.25
```

</details>

### Cloud provider

<details>
GKE
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by msau42** [comment]
@kubernetes/sig-storage-bugs 
/triage accepted

**Comment by msau42** [comment]
One question is why do attach and mount depend on CSINode?

I think in this case, this cluster doesn't have this fix https://github.com/kubernetes/kubernetes/pull/115464. And mount is blocked on attach.

But not having a CSINode object in general also causes scheduling and provisioning issues too.

**Comment by sunnylovestiramisu** [comment]
/assign @sunnylovestiramisu 

Taking a look this week

**Comment by sunnylovestiramisu** [comment]
Discussed two approaches:

1. Make the default deletion to do foreground deletion instead of the background deletion right now. The problem is that there are a few different controllers have already been deleting the node, there will be a lot of work to change all of them to foreground deletion. At the same time, it is difficult to stop new controllers to use foreground deletion by default.

2. [Preferred] Since the Kubelet creates a node object at start up time, and csi node manager will create a csi node object accordingly, we should make the csinode manager component initialization blocking the node manager initialization.

**Comment by msau42** [comment]
I don't think it should block the Node initiatialization, however it should block csi initialization.

**Comment by sunnylovestiramisu** [comment]
Okay so currently we do call [func InstallCSIDriver in the nodeinfomanager](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/nodeinfomanager/nodeinfomanager.go#L109), and then we call [func updateCSINode](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/nodeinfomanager/nodeinfomanager.go#L349), the problem is that we only recreate the CSINode object when it is [not nil](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/nodeinfomanager/nodeinfomanager.go#L382). We can remove the nil check and recreate each time.

**Comment by sunnylovestiramisu** [comment]
@msau42 I am lining towards option 1 because it has simpler code logic even it may do additional update calls. What do you think?

**Comment by sunnylovestiramisu** [comment]
Tentative fix: https://github.com/kubernetes/kubernetes/compare/master...sunnylovestiramisu:kubernetes:fix-csinode

The nodeinfomanager_test.go does not have existing unit test for the func InitializeCSINodeWithAnnotation, need to add extra set up for this.

**Comment by sunnylovestiramisu** [comment]
So the conclusion is that for a short term fix, we should add logic for the following:

1. Pass down [registrationCompleted](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1042) in the Kubelet
2. In the [csi_plugin.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_plugin.go#L277), add logic to wait for registrationCompleted.
3. In the [pkg/volume/csi/nodeinfomanager/nodeinfomanager.go](https://github.com/kubernetes/kubernetes/compare/master...sunnylovestiramisu:kubernetes:fix-csinode), if the CSINode exists but it has different node UID, delete the CSINode and recreate it.


For the long term we should reconcile CSINode object similar to [func syncNodeStatus](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L514).

**Comment by sunnylovestiramisu** [comment]
@bobbypage Can I use Node -> NodeStatus -> NodePhase(NodePhase is the recently observed lifecycle phase of the node) as an evidence of a node is [registrationCompleted](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1042)? Does NodePending indicate the node has been registered?

```
const (
  // NodePending means the node has been created/added by the system, but not configured.
  NodePending NodePhase = "Pending"
  // NodeRunning means the node has been configured and has Kubernetes components running.
  NodeRunning NodePhase = "Running"
  // NodeTerminated means the node has been removed from the cluster.
  NodeTerminated NodePhase = "Terminated"
)
```

**Comment by sunnylovestiramisu** [comment]
In the kubelet we should have a func accessible to the csi_plugin to indicate the registrationCompleted.

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go

**Comment by multani** [comment]
This issue is still occurring (had a case just yesterday).

**Comment by sunnylovestiramisu** [comment]
Have a tentative fix in https://github.com/kubernetes/kubernetes/compare/master...sunnylovestiramisu:kubernetes:fix-csinode

But it is difficult to reproduce and verify.

**Comment by multani** [comment]
@sunnylovestiramisu is there any information that we could provide that would help to identify how the bug happens?

We don't have this happening often nor regularly, but we are seeing it happening about once every weeks. We don't deploy this ourselves (we deploy through GKE) so I won't be able to easily deploy any custom patch ; we can provide logs or other information (possibly), as needed.

**Comment by sunnylovestiramisu** [comment]
@multani If you have a way to reproduce it(even on GKE clusters), that will be helpful, but other than that we kinda know what was happening.

**Comment by gnufied** [comment]
cc @gnufied
